[NativeAOT] Linux/ARM bring-up (4/n)

filipnavara opened the pull request:
Contributes to https://github.com/dotnet/runtimelab/issues/833

Implements thread hijacking, fixes tests with recursive generics.

filipnavara committed:
Fix recursive generics for ARM

filipnavara committed:
Fix compilation on Debian Bookworm

filipnavara committed:
Implement thread return address hijacking for ARM

filipnavara commented on src/coreclr/nativeaot/Runtime/unix/UnixNativeCodeManager.cpp:
```diff
@@ -775,13 +775,25 @@ bool UnixNativeCodeManager::GetReturnAddressHijackInfo(MethodInfo *    pMethodIn
     *pRetValueKind = GetGcRefKind(decoder.GetReturnKind());
 
+#if defined(TARGET_ARM)
+    // FIXME: Figure out how to encode this correctly or implement TrailingEpilogueInstructionsCount
+    if (!decoder.IsInterruptible())
+    {
+        return false;
+    }
```
This needs to be resolved to make this PR ready. We generally have couple of options:
- Implement `TrailingEpilogueInstructionsCount` for ARM/Thumb. That's doable but error-prone.
- Teach JIT to record DWARF epilogs. Seems easy but comes at a size cost. Also, if we move to emitting ARM EHABI for exception handling, this would not work anymore since the generic EHABI codes don't cover unwinding prologs and epilogs.
- Record the size of prolog/epilog (or rather non-prolog/non-epilog ranges) in the JIT into GC info and use it here. This should have less overhead than generic DWARD unwinding and it can even be platform neutral.

jkotas replied:
cc @VSadov

filipnavara replied:
The register corruption was a result of `INLINE_THREAD` macro trashing the `r1` registers. Once I fixed that I got several successful smoke test runs. There's still an occasional failure in the `DynamicGenerics` test. I didn't get to investigate that one since I didn't catch it with debugger yet.

filipnavara committed:
Implement TrailingEpilogueInstructionsCount for ARM

MichalStrehovsky commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/Target_ARM/ARMReadyToRunGenericHelperNode.cs:
```diff
@@ -39,6 +39,18 @@ protected void EmitDictionaryLookup(NodeFactory factory, ref ARMEmitter encoder,
+
+            // If there's any invalid entries, we need to test for them
+            //
+            // Only do this in relocsOnly to make it easier to weed out bugs - the _hasInvalidEntries
```
I know this is copy paste but the comment is wrong. Could you fix the other 2 instances too please?

```suggestion
            // Skip this in relocsOnly to make it easier to weed out bugs - the _hasInvalidEntries
```

filipnavara replied:
Yep, will do.

filipnavara committed:
Fix comment

filipnavara committed:
Fix bugs in RhpGcProbeHijack logic

filipnavara committed:
Fix register trashing by INLINE_GETTHREAD in FixupHijackedCallstack on ARM

filipnavara committed:
Mask the Thumb bit when loading IP from probe frame

filipnavara commented on src/coreclr/nativeaot/Runtime/unix/UnixNativeCodeManager.cpp:
```diff
@@ -367,13 +367,18 @@ bool UnixNativeCodeManager::IsUnwindable(PTR_VOID pvAddress)
 {
     MethodInfo * pMethodInfo = NULL;
 
+#if defined(TARGET_ARM)
+    // FIXME!!!
+    pvAddress = (PTR_VOID)((uintptr_t)pvAddress & ~1);
+#endif
```
TODO: Replace with ASSERT and fix the callers.

filipnavara commented on src/coreclr/nativeaot/Runtime/unix/UnixNativeCodeManager.cpp:
```diff
@@ -781,8 +999,10 @@ bool UnixNativeCodeManager::GetReturnAddressHijackInfo(MethodInfo *    pMethodIn
 
     GcInfoDecoder decoder(GCInfoToken(p), flags);
     *pRetValueKind = GetGcRefKind(decoder.GetReturnKind());
+    // FIXME!!!
+    TADDR ip = PCODEToPINSTR((PCODE)pRegisterSet->IP);
```
TODO: Replace with ASSERT and fix the callers.

filipnavara committed:
Disable DwarfDump on linux-arm

VSadov commented on src/coreclr/nativeaot/Runtime/arm/GcProbe.S:
```diff
@@ -8,6 +8,104 @@
+    // incoming register values into it.
+    PROLOG_STACK_ALLOC  4                   // Space for saved APSR
+    PROLOG_VPUSH        "{d0-d3}"           // Save floating point return registers
+    PROLOG_PUSH         "{r0-r3,lr}"        // Save volatile registers
```
This macro is used to push a frame when a method returns. Volatile regiters are dead at that point, so no need to save. 
Except for the return registers that, although volatile, may contain live values.

VSadov replied:
I think PUSH_PROBE_FRAME should be similar to corresponding PUSH_COOP_PINVOKE_FRAME + the same diff as between arm64 versions of PUSH_COOP_PINVOKE_FRAME and PUSH_PROBE_FRAME - i.e. PUSH_PROBE_FRAME also saves return registers and takes flags parameter.

filipnavara replied:
I wrote it as a bit of a mix between `PUSH_COOP_PINVOKE_FRAME` and the code in GcProbe.S. There are some parts which are unnecessary dead code, especially since it's only ever called with one specific BITMASK parameter.  I intend to clean it up but I wanted to fix the crashes first.

VSadov replied:
One possibly helpful thing could be to implement arm32 version of `ForEachPossibleObjectRef` and try running smoke tests with `export DOTNET_gcConservative=1` - to see how far it can get without involving hijacking.

Some libraries tests would have problems with conservative mode, since they test/rely on deterministic finalization, but smoke tests should generally pass. Unless there are issues besides hijacking. Sometimes it is useful to be sure about that.

filipnavara replied:
I got the hijacking mostly working but I see HUGE retry counts in `SuspendAllThreads`. That happens both on QEMU and real HW. It's slightly less bad when SIGUSR1 is used instead of the real-time SIGRTMIN for the injection.

 The only failing test at the moment is `ThreadLocalStatics.TLSTesting.ThreadLocalStatics_Test` inside `DynamicGenerics`, and it is flaky and passes about 50% of the time on the Raspberry Pi.

VSadov replied:
> HUGE retry counts

Is that because we keep hitting non-unwindable code?

filipnavara replied:
> Volatile regiters are dead at that point, so no need to save.

Fixed this part.

filipnavara replied:
> Is that because we keep hitting non-unwindable code?

I have many different theories but I don't have enough evidence to support them. I don't have reliable data about hitting non-unwindable code but I am reasonably sure that's not the case. There may be an issue with slow unwinding (again, no data to support it, but it's largely unexplored and unprofiled code paths).

On QEMU (ARM user emulation on x64 host) I see a lot of race conditions with the logic in `RhpPInvokeReturn` where `RhpWaitForGC2` is enter and it temporarily removes the `PInvokeTransitionFrame`. Whenever I stop it in the debugger in the `SuspendAllThreads` loop the return value of `CacheTransitionFrameForSuspend` was clearly `NULL` but the frame is present in the thread structure. I thought it's an artifact of how the QEMU emulation works and how it handles signals but I witnessed a very similar pattern on the Raspberry Pi.

filipnavara replied:
Here's one example of the high iteration count in `SuspendAllThreads`: https://gist.github.com/filipnavara/b42ce8459ff36d1d671e89a0b09f1144

(not sure if there's anything relevant there, just posting it for posterity)

filipnavara committed:
Cleanup

filipnavara commented:
I managed to get a stack trace from the last failing smoke test:
```
Process terminated. Assertion failed.
   at System.Diagnostics.DebugProvider.Fail(String, String) + 0x38
   at System.Diagnostics.Debug.Fail(String, String) + 0x56
   at System.Diagnostics.Debug.Assert(Boolean, String, String) + 0x32
   at System.Diagnostics.Debug.Assert(Boolean) + 0x26
   at System.Runtime.TypeCast.IsInstanceOfClass(MethodTable*, Object) + 0x118
   at System.Runtime.TypeCast.IsInstanceOfAny_NoCacheLookup(MethodTable*, Object) + 0xf0
   at System.Runtime.TypeCast.StelemRef_Helper_NoCacheLookup(Object&, MethodTable*, Object) + 0x22
   at System.Runtime.TypeCast.StelemRef_Helper(Object&, MethodTable*, Object) + 0x68
   at System.Runtime.TypeCast.StelemRef(Array, IntPtr, Object) + 0x100
   at System.Collections.Generic.LowLevelList`1.Add(T) + 0x5c
   at System.Reflection.Runtime.General.NamespaceChain..ctor(MetadataReader, NamespaceDefinitionHandle) + 0xd8
   at System.Reflection.Runtime.TypeInfos.NativeFormat.NativeFormatRuntimeNamedTypeInfo.get_NamespaceChain() + 0x130
   at System.Reflection.Runtime.TypeInfos.NativeFormat.NativeFormatRuntimeNamedTypeInfo.get_Namespace() + 0x1e
   at System.Reflection.Runtime.TypeInfos.RuntimeNamedTypeInfo.get_FullName() + 0x12c
   at System.Reflection.Runtime.TypeInfos.NativeFormat.NativeFormatRuntimeNamedTypeInfo.ToString() + 0x25c
   at System.Reflection.Runtime.TypeInfos.RuntimeTypeInfo.EstablishDebugName() + 0x5a
   at System.Reflection.Runtime.TypeInfos.NativeFormat.NativeFormatRuntimeNamedTypeInfo.GetRuntimeNamedTypeInfo(MetadataReader, TypeDefinitionHandle, RuntimeTypeHandle) + 0xe2
   at System.Reflection.Runtime.General.TypeUnifier.GetNamedType(TypeDefinitionHandle, MetadataReader, RuntimeTypeHandle) + 0x20
   at System.Reflection.Runtime.General.TypeUnifier.GetNamedType(TypeDefinitionHandle, MetadataReader) + 0x24
   at System.Reflection.Runtime.General.TypeResolver.ResolveTypeDefinition(TypeDefinitionHandle, MetadataReader) + 0x1e
   at System.Reflection.Runtime.General.TypeResolver.TryResolve(Handle, MetadataReader, TypeContext, Exception&) + 0x60
   at System.Reflection.Runtime.General.QSignatureTypeHandle.TryResolve(TypeContext, Exception&) + 0x7e
   at System.Reflection.Runtime.General.QSignatureTypeHandle.Resolve(TypeContext) + 0x2a
   at System.Reflection.Runtime.ParameterInfos.RuntimeMethodParameterInfo.get_ParameterType() + 0x52
   at System.Reflection.Runtime.MethodInfos.RuntimeMethodInfo.get_ReturnType() + 0x40
   at System.Reflection.DynamicInvokeInfo..ctor(MethodBase, IntPtr) + 0x33a
   at Internal.Reflection.Execution.MethodInvokeInfo..ctor(MethodBase, IntPtr) + 0x1a
   at Internal.Reflection.Execution.ExecutionEnvironmentImplementation.TryGetMethodInvokeInfo(RuntimeTypeHandle, QMethodDefinition, RuntimeTypeHandle[], MethodBase, MethodSignatureComparer&, CanonicalFormKind) + 0x12c
   at Internal.Reflection.Execution.ExecutionEnvironmentImplementation.TryGetMethodInvoker(RuntimeTypeHandle, QMethodDefinition, RuntimeTypeHandle[]) + 0x96
   at Internal.Reflection.Core.Execution.ExecutionEnvironment.GetMethodInvoker(RuntimeTypeInfo, QMethodDefinition, RuntimeTypeInfo[], MemberInfo, Exception&) + 0x19c
   at System.Reflection.Runtime.MethodInfos.NativeFormat.NativeFormatMethodCommon.GetUncachedMethodInvoker(RuntimeTypeInfo[], MemberInfo, Exception&) + 0x6e
   at System.Reflection.Runtime.MethodInfos.RuntimeNamedMethodInfo`1.GetUncachedMethodInvoker(RuntimeTypeInfo[], MemberInfo) + 0x30
   at System.Reflection.Runtime.MethodInfos.RuntimeNamedMethodInfo`1.get_UncachedMethodInvoker() + 0x74
   at System.Reflection.Runtime.MethodInfos.RuntimeMethodInfo.get_MethodInvoker() + 0x4a
   at System.Reflection.Runtime.MethodInfos.RuntimeMethodInfo.Invoke(Object, BindingFlags, Binder, Object[], CultureInfo) + 0x28
   at System.Reflection.MethodBase.Invoke(Object, Object[]) + 0x3a
   at ThreadLocalStatics.TLSTesting.MakeType2(Type, Type, Boolean) + 0xa36
   at ThreadLocalStatics.TLSTesting.<>c__DisplayClass3_0.<MultiThreaded_Test>b__0() + 0x50
   at System.Threading.Tasks.Task.InnerInvoke() + 0x66
   at System.Threading.Tasks.Task.<>c.<.cctor>b__281_0(Object obj) + 0x4c
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread, ExecutionContext, ContextCallback, Object) + 0x98
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task&, Thread) + 0x226
   at System.Threading.Tasks.Task.ExecuteEntryUnsafe(Thread) + 0x72
   at System.Threading.Tasks.Task.ExecuteFromThreadPool(Thread) + 0x14
   at System.Threading.ThreadPoolWorkQueue.DispatchWorkItem(Object, Thread) + 0x54
   at System.Threading.ThreadPoolWorkQueue.Dispatch() + 0x3c8
   at System.Threading.PortableThreadPool.WorkerThread.WorkerDoWork(PortableThreadPool, Boolean&) + 0x3c
   at System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart() + 0x23c
   at DynamicGenerics!<BaseAddress>+0x80aa6d
   at System.Threading.Thread.StartHelper.RunWorker() + 0x5e
   at System.Threading.Thread.StartHelper.Run() + 0x90
   at System.Threading.Thread.StartThread(IntPtr) + 0xc8
   at System.Threading.Thread.ThreadEntryPoint(IntPtr) + 0x1a
```

It is flaky though, and it may be failing with other errors at different times.

filipnavara commented:
Just had another failed test run, this time two test failed:

```
--------------------------------
Running Test: ThreadLocalStatics.TLSTesting.ThreadLocalStatics_Test
--------------------------------------------------
Debug Assertion Violation

Expression: 'pvRetAddr != NULL'

File: /home/navara/runtime/src/coreclr/nativeaot/Runtime/thread.cpp, Line: 813
--------------------------------------------------
Expected: 100
Actual: 134
END EXECUTION - FAILED
Test Harness Exitcode is : 1
To run the test:
> set CORE_ROOT=/home/navara/runtime/artifacts/tests/coreclr/linux.arm.Debug/Tests/Core_Root
> /home/navara/runtime/artifacts/tests/coreclr/linux.arm.Debug/nativeaot/SmokeTests/DynamicGenerics/DynamicGenerics/DynamicGenerics.sh


Failed test: nativeaot_SmokeTests::_UnitTests_UnitTests_UnitTests_sh (nativeaot/SmokeTests/UnitTests/UnitTests/UnitTests.sh) (nativeaot.SmokeTests.XUnitWrapper.dll)
Process terminated. Assertion failed.
   at System.Diagnostics.DebugProvider.Fail(String, String) + 0x38
   at System.Diagnostics.Debug.Fail(String, String) + 0x56
   at System.Diagnostics.Debug.Assert(Boolean, String, String) + 0x32
   at System.Diagnostics.Debug.Assert(Boolean) + 0x26
   at System.Threading.ObjectHeader.SetSyncEntryIndex(Int32*, Int32) + 0x2c
   at System.Threading.SyncTable.AssignEntry(Object, Int32*) + 0x250
   at System.Threading.ObjectHeader.GetSyncIndex(Object) + 0x68
   at System.Threading.ObjectHeader.GetLockObject(Object) + 0x1a
   at System.Threading.Monitor.<>c.<.cctor>b__22_0(Object o) + 0x1a
   at System.Runtime.CompilerServices.ConditionalWeakTable`2.GetValueLocked(TKey, ConditionalWeakTable`2.CreateValueCallback) + 0x36
   at System.Runtime.CompilerServices.ConditionalWeakTable`2.GetValue(TKey, ConditionalWeakTable`2.CreateValueCallback) + 0x4a
   at System.Threading.Monitor.GetCondition(Object) + 0xb8
   at System.Threading.Monitor.Wait(Object, Int32) + 0x1a
   at System.Threading.ManualResetEventSlim.Wait(Int32, CancellationToken) + 0x282
   at System.Threading.ManualResetEventSlim.Wait() + 0x2e
   at ThreadTest.<>c__DisplayClass10_0.<TestConcurrentIsBackgroundProperty>b__0() + 0x4a
   at System.Threading.Thread.StartHelper.RunWorker() + 0x5e
   at System.Threading.Thread.StartHelper.Run() + 0x90
   at System.Threading.Thread.StartThread(IntPtr) + 0xc8
   at System.Threading.Thread.ThreadEntryPoint(IntPtr) + 0x1a
```

I have seen this before. The `pvRetAddr != NULL` error happens if the unwinding in the hijack callback doesn't detect the epilog. I thought I got rid of it by clearing the thumb bit in the IP which mistakenly made the prolog/epilog detection check unaligned instruction stream. In the last commit I replaced it with `ASSERT` because I was convinced that I fixed the root cause (`PUSH_PROBE_FRAME` pushes the `lr` value into the frame which the `StackFrameIterator` then treats as IP, so the bit has to be cleared there).

VSadov commented on src/coreclr/nativeaot/Runtime/arm/GcProbe.S:
```diff
@@ -8,6 +8,100 @@
+.macro PUSH_PROBE_FRAME threadReg, trashReg, BITMASK
+    // Define the method prolog, allocating enough stack space for the PInvokeTransitionFrame and saving
+    // incoming register values into it.
+    PROLOG_PUSH         "{r0-r1}"           // Save return registers
```
Do we need to push down SP and store sp/fp first?

filipnavara replied:
I don't think it's necessary. Other archs don't do it either. They just construct the `PInvokeTransitionFrame` on the stack and make its IP/SP/FP point to the user code before the hijack. That way the `StackFrameIterator` on the GC thread begins unwinding from the correct location. On the hijack thread this is just pushed, then waits for end of GC, and pops it in the reverse order.

VSadov replied:
I am not sure if Linux/ARM32 has a sufficient red zone. I can't find any good docs about that. 
Even if there is, perhaps better not rely on it as it might be OS-specific.

filipnavara replied:
I'm not sure I fully understand your concern here. The ARM/Thumb instruction set is not really well tailored for the `add sp, #imm` pattern and then accessing the stack by relative writes. That's why it's not done in prologs, or `PUSH_COOP_PINVOKE_FRAME`.

VSadov replied:
Ah, I misread this as if we possibly use the stack beyond SP (that could cause problems with signal handlers, in theory). But we are not, so this should be ok.

VSadov commented:
For the write barriers, you may want to try disabling the following for starters - at least to see if that accounts for all the remaining issues.  

```
  add_definitions(-DFEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP)
  add_definitions(-DFEATURE_MANUALLY_MANAGED_CARD_BUNDLES)
```

It looks like the ARM32 write barriers do not have support for that. I am not sure if that works on 32bit at all.

filipnavara commented:
Thanks for the debugging tips. I didn't do enough runs to verify or confirm it, but returning `false` from `GetReturnAddressHijackInfo` made the tests pass several times. That said, I think this PR is improvement over the status quo and I would continue the work/investigation in a next one.

(I intend to fix the DWARF info not to contain the Thumb bit in the addresses and align correctly with the instruction boundaries, but it requires fixing up the managed stack trace logic to match it.)

filipnavara committed:
Emit DWARF info with instruction addresses without Thumb bit (matches clang)

filipnavara commented:
I finally managed to catch the `DynamicGenerics` crash in the debugger and it's actually quite fun one.

Let's start with the back trace:
```
* thread #11, name = 'DynamicGenerics', stop reason = signal SIGSEGV: invalid address (fault address: 0xc100ae9c)
  * frame #0: 0xc100ae9c
    frame #1: 0x00ad0c1e DynamicGenerics`S_P_TypeLoader_Internal_TypeSystem_TypeDescExtensions__IsWellKnownType(type=0xb3e641a0, wellKnownType=String) at TypeSystemExtensions.cs:27
    frame #2: 0x00aca65a DynamicGenerics`S_P_TypeLoader_Internal_TypeSystem_TypeDesc__get_IsString(this=0xb3e641a0) at TypeDesc.cs:273
```

The instruction on frame 1 is `blx r3`. So, let's have a look at the registers:
```
General Purpose Registers:
        r0 = 0xb3e9fe5c
        r1 = 0x00000014
        r2 = 0x00000000
        r3 = 0x0082d1a1  DynamicGenerics`__VirtualCall_S_P_TypeLoader_Internal_TypeSystem_TypeSystemContext__GetWellKnownType + 1
        r4 = 0x00000000
        r5 = 0x00000000
        r6 = 0xad0fbe08
        r7 = 0x000000af
        r8 = 0xac8fd000
        r9 = 0xb0ade78f
       r10 = 0x00000000
       r11 = 0xad0fbd08
       r12 = 0xc100ae9c
        sp = 0xad0fbcd8
        lr = 0x00ad0c1f  DynamicGenerics`S_P_TypeLoader_Internal_TypeSystem_TypeDescExtensions__IsWellKnownType + 78 at TypeSystemExtensions.cs:27
        pc = 0xc100ae9c
      cpsr = 0x60000010
```

`$r3` is still valid and pointing to `__VirtualCall_S_P_TypeLoader_Internal_TypeSystem_TypeSystemContext__GetWellKnownType`, so presumably something happened inside the `__VirtualCall` method. That's quite a big clue since that method has no prolog/epilog.

The faulting address is `0xc100ae9c` and it's in $r12. That's not surprising either, since the `VirtualCall` helper uses `$r12` as the scratch register to compute the jump. Curiously, the address looks like shifted by one byte from something that looks like a valid destination of the virtual call:

```
(lldb) image lookup -va 0x00ae9c32
      Address: DynamicGenerics[0x006e9c32] (DynamicGenerics.PT_LOAD[1].__managedcode + 2232674)
      Summary: DynamicGenerics`S_P_TypeLoader_Internal_Runtime_TypeLoader_TypeLoaderTypeSystemContext__GetWellKnownType + 1 at TypeLoaderTypeSystemContext.cs:52
```

filipnavara committed:
Report R2/R3 registers in ForEachPossibleObjectRef

filipnavara commented:
I captured few more of the crashes. Seems like they don't happen with `DOTNET_gcConservative=1`. There are already trash values on a stack if I follow the code back a bit... so that goes along with the GC hole theory.

filipnavara committed:
Ensure that PInvokeTransitionFrame(s) on the stack are 8-byte aligned. Save FP return values on hijack.

filipnavara committed:
Tame the Thumb bit

filipnavara committed:
Fix GC hole when thread hijack happens with r0 register holding a reference (eg. boxed int)

filipnavara commented:
I managed to find the GC hole. Fix is committed, rebuilding and rerunning tests now.

filipnavara commented:
The tests pass on Raspberry Pi now:

```
Time [secs] | Total | Passed | Failed | Skipped | Assembly Execution Summary
============================================================================
      0.522 |     1 |      1 |      0 |       0 | nativeaot.CustomMain.XUnitWrapper.dll
      0.211 |     1 |      1 |      0 |       0 | nativeaot.GenerateUnmanagedEntryPoints.XUnitWrapper.dll
     44.544 |    14 |     14 |      0 |       0 | nativeaot.SmokeTests.XUnitWrapper.dll
      0.194 |     1 |      1 |      0 |       0 | nativeaot.StartupHook.XUnitWrapper.dll
----------------------------------------------------------------------------
     45.471 |    17 |     17 |      0 |       0 | (total)
```

Postmortem:

The main bug was that thread hijacking failed to report the method return value in R0 register as GC reference. It was reproducible with the TLS test in DynamicGenerics. I enabled `LF_GC | LF_GCROOTS` logging and captured a crash with a log and a crash dump. Then I found the last stack walk on the thread with trashed values and after some poking it became obvious what went wrong.

Second bug is that the Thumb bit was often not masked out from IP during stack walks. The LR register (or custom frames) saved on the stack have the Thumb bit set. It needs to be cleared before doing the GC enumeration to ensure that [subtracting -1 here actually gets to previous instruction](https://github.com/dotnet/runtime/blob/bd450af318265edb597c738cf81b0a6410d8aa33/src/coreclr/nativeaot/Runtime/unix/UnixNativeCodeManager.cpp#L208-L217).

jkotas reviewed (approved):
Looks great to me. Thank you!

@VSadov Could you please sign-off on this change as well?

VSadov commented on src/coreclr/nativeaot/Runtime/arm/GcProbe.S:
```diff
@@ -8,6 +8,105 @@
+    // incoming register values into it.
+    PROLOG_VPUSH        "{d0-d3}"           // Save d0-d3 which can have the floating point return value
+    PROLOG_STACK_ALLOC  4                   // Padding for 8-byte alignment
+    PROLOG_PUSH         "{r0-r1}"           // Save return registers
```
```suggestion
    PROLOG_PUSH         "{r0,r1}"           // Save return registers
```

VSadov reviewed (approved):
Very nice! Thank you!!

VSadov commented:
Couple things that do not need to go in this change (perhaps better be logged as issues to follow up):

- I am not sure if background GC can be supported on 32bit, but the current state of GC barriers on ARM32 does not support it, so it should be disabled. Disabling write watch will do that. 
We need to disable it here, similarly to card bundles.
https://github.com/dotnet/runtime/blob/ffc1174638f0dd6ebba08671e931e1fce32bf4d1/src/coreclr/nativeaot/Runtime/CMakeLists.txt#L238

Maybe it gets disabled somehow anyways, but it seems better to be explicit about it.

- FEATURE_64BIT_ALIGNMENT may need to be reviewed
It is enabled for ARM32 now
https://github.com/dotnet/runtime/blob/ffc1174638f0dd6ebba08671e931e1fce32bf4d1/src/coreclr/nativeaot/Runtime/CMakeLists.txt#L243

I think there could be some subtle NYIs around that, for example: 
https://github.com/dotnet/runtime/blob/ffc1174638f0dd6ebba08671e931e1fce32bf4d1/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ILScanner.cs#L760

VSadov committed:
Update src/coreclr/nativeaot/Runtime/arm/GcProbe.S
